### ISSUE TITLE
Lorentz Linear result outside the manifold

### ISSUE THREAD

**Issue by inboxedshoe** [chenweize1998/fully-hyperbolic-nn#3]
I've been trying to implement a 2d convolution like operation using this method, I get the data and project it to the manifold, do a few steps and then pass them through the lorentz layer. 
Just before I pass the data in, I perform the "check_on_manifold" function and it returns True. However, right after the layer it becomes False. Is there any specific reason for that? Should I reproject after using the layer?

**Comment by chenweize1998** [comment]
I don't think this should happen. Could you paste the input data here so I can look into it? Or could you give me a code to re-produce?

**Comment by inboxedshoe** [comment]
Yeah definitely,

```
# create fake data
batch_size = 1
channels = 5
h, w = 12, 12
image = torch.rand(batch_size, channels, h, w)  # input image

man = Lorentz()

# initialize operations and layers
unfold = torch.nn.Unfold(kernel_size=(3, 3), dilation=1, stride=2, padding=3)
LL1 = LorentzLinear(3*3, 10, manifold=man, bias = True, dropout=0).to(device="cuda:0")

# project onto the manifold
hyper_image = man.projx(image).to(device="cuda:0")

# project onto tangent space so we can do euclidean operations normally
x = man.logmap0(hyper_image)

# unfold and reshape
patches = unfold(x)
patches = patches.view(batch_size, -1, channels, 3*3)  

# project back to the manifold
patches = man.expmap0(patches)

# upto here testing patches will result in (True, None)
print(man._check_point_on_manifold(patches))

# testing patches after layer here will fail
patches = LL1(patches)

print(man._check_point_on_manifold(patches))
```

The resulting failure is "(False, "'x' minkowski quadratic form is not equal to -1.0")".

I did some more testing and I found that if I set the padding to 0, I get no issues. It might be the extra 0s but before passing it in, the patches are still on the manifold.

**Comment by inboxedshoe** [comment]
I did some more tests on simpler scenarios, stacking Lorentz layers also seems to have the same effect. Tried the same thing with mt.onmt versions of the manifold and the linear layer

```
from gcn.layers.hyp_layers import LorentzLinear
from gcn.manifolds.lorentz import Lorentz
import torch


manifold = Lorentz()

layer1 = LorentzLinear(5, 20, manifold=manifold)
layer2 = LorentzLinear(20, 1, manifold=manifold)

x = manifold.projx(torch.rand((5, 5)))

x = layer1(x)
print(manifold._check_point_on_manifold(x))

x = layer2(x)
print(manifold._check_point_on_manifold(x))

```

**Comment by chenweize1998** [comment]
Thanks for the scripts! Yeah, I can reproduce the unexpected result. For the first script, the out-of-manifold points come from the origin of the Lorentz model (1.0, 0, 0, ...). Our code does not handle this boundary case. The solution can be checking this boundary case before output
```python
def forward(self, x):
    if self.nonlin is not None:
        x = self.nonlin(x)
    x = self.weight(self.dropout(x))
    x_narrow = x.narrow(-1, 1, x.shape[-1] - 1)
    time = x.narrow(-1, 0, 1).sigmoid() * self.scale.exp() + 1.1
    sum_square = (x_narrow * x_narrow).sum(dim=-1, keepdim=True)
    scale = (time * time - 1) / \
        sum_square.clamp_min(1e-8)
    time = time.masked_fill(sum_square==0, 1)
    x = torch.cat([time, x_narrow * scale.sqrt()], dim=-1)
    return x
```

As for the second script, the point is transformed into a 1-dimensional point in the second linear layer. However, the Lorentz model does not lie in the 1-dimensional space. Simply speaking, for any point x in the Lorentz model, the Lorentzian inner product between x and itself should be -1. And if x is a 1-dimensional point, then according to the definition, we should have -x * x = -1. We generally not care about this case, since x can only be -1 or 1. Therefore, it makes no sense to transform to a "1-dimensional Lorenz model".

**Comment by inboxedshoe** [comment]
Ahhh, I'm still learning the mathematics of the model but that explains things clearly. I can confirm this works now. Thank you very much for the time and answers!

**Comment by inboxedshoe** [comment]
Heyy, I was trying to tune the K value for the manifold when I noticed that any k other than 1 fails the check again. I did a small change to the code based on my understanding of the paper and got 
```python
def forward(self, x):
    if self.nonlin is not None:
        x = self.nonlin(x)
    x = self.weight(self.dropout(x))
    x_narrow = x.narrow(-1, 1, x.shape[-1] - 1)
    time = x.narrow(-1, 0, 1).sigmoid() * self.scale.exp() + 1.1
    sum_square = (x_narrow * x_narrow).sum(dim=-1, keepdim=True)
    scale = (time * time - 1 * self.manifold.k) / \
        sum_square.clamp_min(1e-8)
    time = time.masked_fill(sum_square==0, 1)
    x = torch.cat([time, x_narrow * scale.sqrt()], dim=-1)
    return x
```
the output is now sometimes on the manifold and sometimes not, any idea? Also, after a second pass-through, I looked at the output from this snippet of the check

```python
        quad_form = -x.narrow(dim, 0, 1) + x.narrow(dim, 1, dn).sum(
            dim=dim, keepdim=True
        )
```

it looks like 

```python
tensor([[[-37.0527],
         [ -1.1500],
         [ -1.1500],
         ...,
         [ -1.1500],
         [ -1.1500],
         [ -1.1500]],
        [[-37.0527],
.........
```
Here I'm using a k value of 1.15 so it appears that all the values are correct except for element 0 of the last dim.

**Comment by chenweize1998** [comment]
Sadly I could not reproduce the result. Could you give me the input that cause the problem? However, the code does have some potential bugs:
```python
time = x.narrow(-1, 0, 1).sigmoid() * self.scale.exp() + 1.1
```
Here, we brutely add 1.1 to ensure the calculated time axis is bigger than 1. You should Change this 1.1 to any other value greater than sqrt(k). But since sqrt(1.15) < 1.1, this seems to not be the cause this time. Nevertheless, it should be changed if you want to work with different k. To see why, we still need to start from the definition of the Lorentz model. Let x be a point on the Lorentz model, and let x_i be the i-th axis of x, then according to the definition, we have -x_0^2+x_1^2+x_2^2+...=-k. Only when x_1=x_2=0, x_0 obtains the minimum value sqrt(k) (We only care about the upper sheet of the Lorentz model). That is, any point in the Lorentz model should have a time axis x_0 greater than sqrt(k). When k=1, x_0>1. Therefore, we add 1.1 in our code to ensure the range of the calculated time is bigger than 1.

Another possible cause is
```python
time = time.masked_fill(sum_square==0, 1)
```
for k != 1, it should be 
```python
time = time.masked_fill(sum_square==0, sqrt(self.manifold.k))
```

If the problem still exists, I will take a closer look after you provide the input.

**Comment by chenweize1998** [comment]
Also, if you want to work with k!=1, the lmath.py should also be changed. Our lmath.py is adapted from https://github.com/geoopt/geoopt/blob/master/geoopt/manifolds/lorentz/math.py. Using this original script should work fine I guess.

**Comment by inboxedshoe** [comment]
So so far I tried adding both fixes and switching to the original geoopt math file without success. I attached an example of a problem tensor that causes the issue (zipped pytorch pt file). 

Also thanks a lot for the explanations they tend to make things a lot clearer at least from an inner working perspective. 

[problem_tensor.zip](https://github.com/chenweize1998/fully-hyperbolic-nn/files/8812871/problem_tensor.zip)

**Comment by chenweize1998** [comment]
Ah, boundary case once again. Change the boundary check and it should be corrected now.
```python
def forward(self, x):
    if self.nonlin is not None:
        x = self.nonlin(x)
    x = self.weight(self.dropout(x))
    x_narrow = x.narrow(-1, 1, x.shape[-1] - 1)
    time = x.narrow(-1, 0, 1).sigmoid() * self.scale.exp() + self.manifold.k.sqrt() + 1e-2
    sum_square = (x_narrow * x_narrow).sum(dim=-1, keepdim=True)
    scale = (time * time - self.manifold.k) / \
        sum_square.clamp_min(1e-8)
    mask = sum_square <= 1e-8
    time = time.masked_fill(mask, self.manifold.k.sqrt())
    scale = scale.masked_fill(mask, 0)
    x = torch.cat([time, x_narrow * scale.sqrt()], dim=-1)
    return x
```

Due to the numerical precision, `_check_point_on_manifold` method still sometimes returns False on the output. You can either set the initialization value of the `scale` in LorentzLinear to a smaller value, or use double precision for the calculation.

**Comment by inboxedshoe** [comment]
Yesss, I can confirm this works now (for the rest of the model as well). I'm currently using float64 anyways for the testing phases atleast. Though the huge memory cost is annoying. 

Thank you very much for the help again!
